Themis/ViewModels/Assessment/AssessmentViewModel.swift

```diff
@@ -8,37 +8,132 @@ class AssessmentViewModel: ObservableObject {
     @Published var submission: BaseSubmission?
     /// Is set automatically when the submission is set
     @Published var participation: BaseParticipation?
-    @Published var assessmentResult = AssessmentResult()
-    @Published var showSubmission = false
+    @Published var assessmentResult: AssessmentResult
     @Published var readOnly: Bool
     @Published var loading = false
     @Published var error: Error?
+    @Published var pencilMode = true
+    @Published var fontSize: CGFloat = 16.0
+    
+    var submissionId: Int?
+    var participationId: Int?
+    var resultId: Int?
+    var exercise: Exercise
     
     private var cancellables: [AnyCancellable] = []
-
-    init(readOnly: Bool) {
+    
+    init(exercise: Exercise, submissionId: Int? = nil, participationId: Int? = nil, resultId: Int? = nil, readOnly: Bool) {
+        self.exercise = exercise
+        self.submissionId = submissionId
+        self.participationId = participationId
+        self.resultId = resultId
+        self.assessmentResult = AssessmentResultFactory.assessmentResult(for: exercise, resultIdFromServer: resultId)
         self.readOnly = readOnly
         
         $submission
             .sink(receiveValue: { self.participation = $0?.participation?.baseParticipation })
             .store(in: &cancellables)
+        
+        $participation
+            .sink(receiveValue: { self.resultId = $0?.results?.last?.id })
+            .store(in: &cancellables)
+    }
+    
+    // Resets some properties of this viewmodel to prepare it for reuse between assessments
+    func resetForNewAssessment() {
+        self.submission = nil
+        self.participation = nil
+        self.loading = false
+        self.error = nil
+        self.submissionId = nil
+        self.participationId = nil
+        self.resultId = nil
+    }
+    
+    /// Resets some properties of this viewmodel that affect the toolbar
+    func resetToolbarProperties() {
+        pencilMode = true
+        fontSize = 16.0
     }
     
     var gradingCriteria: [GradingCriterion] {
         participation?.getExercise()?.gradingCriteria ?? []
     }
+    
+    @MainActor
+    func initSubmission() async {
+        guard submission == nil else {
+            return
+        }
+        
+        switch exercise {
+        case .programming(exercise: _):
+            if readOnly {
+                if let participationId {
```

Comment:
What happens for `readOnly = true, participationId = nil`? Should it throw?

Reply:
Good catch!